File: Labo1/PartieB/hill_climbing_main.py
import random
import copy
import time
import sys
import getopt
import math

MAX_ITER = 200
nb_iter = 0

inputFile = "input/input10.txt"

def loadGrid(src):
    matrix=[]
    f = open(src, 'r')
    for line in f.readlines():
        if(line=='\n'):
            break
        matrix.append([int(x) for x in list(line.replace(".", "0"))[:-1]])
        

    f.close()
    return matrix

def heuristic(grid):
    #Here heuristic is the number of conflicts and we want to minimize it

    conflicts = 0
    # Row: No need to count conficts on rows because the elements are unique on rows

    # Column
    for j in range(9):
        col = [grid[i][j] for i in range(9)]
        conflicts += len(set(col)) - 9
    # Square 3x3
    for x in range(0, 9, 3):
        for y in range(0, 9, 3):
            sub_grid = []
            for i in range(3):
                for j in range(3):
                    sub_grid.append(grid[x+i][y+j])
            conflicts += len(set(sub_grid)) - 9

    return -conflicts


def fillInitialGrid(grid):
    #This function fills the free boxes in each row with numbers

    grid = copy.deepcopy(grid)
    for row in range(9):

        availableNumbers = list(range(1, 10))
        
        #Randomize the order of the numbers
        random.shuffle(availableNumbers)

        for val in grid[row]:
            if(val!=0):
                availableNumbers.remove(val)

        for col in range(9):
            if(not isFixed(col, row)):
                grid[row][col] = availableNumbers[0]
                availableNumbers.pop(0)
    
    return grid

def printgrid(grid):
    for line in grid:
        string = ""
        for number in line:
            string += str(number) + " "
        print(string)
    print()

def isFixed(x, y):
    return initialGrid[y][x] != 0

def swap(grid, x1, y1, x2, y2):
    tmp = grid[y2][x2]
    grid[y2][x2] = grid[y1][x1]
    grid[y1][x1] = tmp


def getNeighbor(grid):
    #we choose a random row, and we chose two random unfixed boxes to swap
    tmp = copy.deepcopy(grid)

    row = random.randint(0, 8)
    x1 = random.randint(0, 8)
    x2 = random.randint(0, 8)

    while(isFixed(x1, row) or isFixed(x2, row) or x1 == x2):
        row = random.randint(0, 8)
        x1 = random.randint(0, 8)
        x2 = random.randint(0, 8)

    swap(tmp, x1, row, x2, row)

    return tmp


def HillClimb(grid):
    
    global nb_iter
    max_iter = MAX_ITER
    
    while(True):
        i=0
        currentGrid = fillInitialGrid(grid)
        currentHeuristic = heuristic(grid)
        

        #round of guessing, we consider its stuck if the heuristic does not improve in max_iter iterations
        while(i < max_iter):

            newGrid = getNeighbor(currentGrid)
            newHeuristic = heuristic(newGrid)

            #if solution, we return it
            if(newHeuristic == 0):
                return newGrid
            
            #if the neighbor is better, it becomes the new grid
            if(newHeuristic < currentHeuristic):
                i=0
                currentGrid = newGrid
                currentHeuristic = newHeuristic

            else:
                i+=1

            nb_iter += 1



# MAIN CALLS
print("Loading grid from file '" + inputFile + "' ...")
initialGrid = loadGrid(inputFile)

print("Running Hill Climbing algorithm...")
startDate = time.time()
result = HillClimb(initialGrid)
executionTime = time.time() - startDate

print("Found a solution :")
printgrid(result)
print(f'Execution time : {math.floor(executionTime)} seconds.')
print(f'Number of iterations : {nb_iter}.')

File: Labo1/PartieB/hill_climbing_main.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>
#include <getopt.h>
#include <pthread.h>

#define SIZE 9
#define NUM_THREADS 8
#define MAX_ITER 200

typedef struct {
    int grid[SIZE][SIZE];
    int thread_id;
} ThreadData;

int initialGrid[SIZE][SIZE];
int shuffle = 1;

int solutionThreadId = 0;
int solution_found = 0;


void initializeRandomSeed() {
    srand(time(NULL) ^ getpid());
}

int loadGrid(const char *src, int matrix[SIZE][SIZE]) {
    FILE *f = fopen(src, "r+");
    if (!f) {
        return 1;
    }

    char line[SIZE + 2];
    int i = 0;
    while (fgets(line, sizeof(line), f)) {
        if (line[0] == '\n') break;
        for (int j = 0; j < SIZE; j++) {
            matrix[i][j] = (line[j] == '.') ? 0 : line[j] - '0';
        }
        i++;
    }

    fclose(f);
    return 0;
}

int heuristic(int grid[SIZE][SIZE]) {
    int conflicts = 0;

    for (int j = 0; j < SIZE; j++) {
        int col[SIZE] = {0};
        for (int i = 0; i < SIZE; i++) {
            col[i] = grid[i][j];
        }
        int unique_count = 0;
        for (int i = 0; i < SIZE; i++) {
            for (int k = i + 1; k < SIZE; k++) {
                if (col[i] == col[k] && col[i] != 0) {
                    unique_count++;
                }
            }
        }
        conflicts += unique_count;
    }

    for (int x = 0; x < SIZE; x += 3) {
        for (int y = 0; y < SIZE; y += 3) {
            int sub_grid[SIZE] = {0};
            int index = 0;
            for (int i = 0; i < 3; i++) {
                for (int j = 0; j < 3; j++) {
                    sub_grid[index++] = grid[x + i][y + j];
                }
            }
            int unique_count = 0;
            for (int i = 0; i < SIZE; i++) {
                for (int k = i + 1; k < SIZE; k++) {
                    if (sub_grid[i] == sub_grid[k] && sub_grid[i] != 0) {
                        unique_count++;
                    }
                }
            }
            conflicts += unique_count;
        }
    }

    return conflicts;
}

void fillInitialGrid(int grid[SIZE][SIZE]) {
    int tempGrid[SIZE][SIZE];
    memcpy(tempGrid, grid, sizeof(tempGrid));

    for (int row = 0; row < SIZE; row++) {
        int availableNumbers[SIZE];
        for (int i = 0; i < SIZE; i++) {
            availableNumbers[i] = i + 1;
        }

        if (shuffle) {
            for (int i = SIZE - 1; i > 0; i--) {
                int j = rand() % (i + 1);
                int temp = availableNumbers[i];
                availableNumbers[i] = availableNumbers[j];
                availableNumbers[j] = temp;
            }
        }

        for (int col = 0; col < SIZE; col++) {
            if (tempGrid[row][col] != 0) {
                for (int i = 0; i < SIZE; i++) {
                    if (availableNumbers[i] == tempGrid[row][col]) {
                        availableNumbers[i] = 0;
                        break;
                    }
                }
            }
        }

        int index = 0;
        for (int col = 0; col < SIZE; col++) {
            if (tempGrid[row][col] == 0) {
                while (availableNumbers[index] == 0) {
                    index++;
                }
                tempGrid[row][col] = availableNumbers[index];
                index++;
            }
        }
    }

    memcpy(grid, tempGrid, sizeof(tempGrid));
}

void printGrid(int grid[SIZE][SIZE]) {
    for (int i = 0; i < SIZE; i++) {
        for (int j = 0; j < SIZE; j++) {
            printf("%d ", grid[i][j]);
        }
        printf("\n");
    }
    printf("\n");
}

int isFixed(int x, int y) {
    return initialGrid[y][x] != 0;
}

void swap(int grid[SIZE][SIZE], int x1, int y1, int x2, int y2) {
    int tmp = grid[y2][x2];
    grid[y2][x2] = grid[y1][x1];
    grid[y1][x1] = tmp;
}

int countFixed(int rowIndex) {
    int count = 0;
    for (int i = 0; i < SIZE; i++) {
        if (isFixed(i, rowIndex)) {
            count++;
        }
    }
    return count;
}

void getNeighbor(int grid[SIZE][SIZE], int result[SIZE][SIZE]) {
    int row, x1, x2;;
    
    do {
        row = rand() % SIZE;
        x1 = rand() % SIZE;
        x2 = rand() % SIZE;
    } while (isFixed(x1, row) || isFixed(x2, row) || x1==x2);
    
    memcpy(result, grid, sizeof(int)*SIZE*SIZE);
    swap(result, x1, row, x2, row);
}

int HillClimb(int grid[SIZE][SIZE]) {
    int currentGrid[SIZE][SIZE];
    int newGrid[SIZE][SIZE];
    int currentHeuristic, newHeuristic;

    while (!solution_found) {

        int i = 0;
        fillInitialGrid(grid);
        memcpy(currentGrid, grid, sizeof(currentGrid));
        currentHeuristic = heuristic(currentGrid);

        while (i < MAX_ITER) {
            getNeighbor(currentGrid, newGrid);
            newHeuristic = heuristic(newGrid);

            if (newHeuristic == 0) {
                memcpy(grid, newGrid, sizeof(newGrid));
                solution_found = 1;
                return 1;
            }

            if (newHeuristic < currentHeuristic) {
                i=0;
                memcpy(currentGrid, newGrid, sizeof(newGrid));
                currentHeuristic = newHeuristic;
            }else{
                i++;
            }
        }
    }

    return 0;
}

void* thread(void* arg) {
    ThreadData* data = (ThreadData*)arg;

    if(HillClimb(data->grid)){
        solutionThreadId = data->thread_id;
    }

    return NULL;
}

int main(int argc, char *argv[]) {
    pthread_t threads[NUM_THREADS];
    ThreadData threadData[NUM_THREADS];

    if(argc == 1){
        printf("Usage: %s [file]\n", argv[0]);
        return 1;
    }

    initializeRandomSeed();
    
    printf("Loading grid from file '%s' ...\n", argv[1]);

    if(loadGrid(argv[1], initialGrid)){
        printf("File does not exists.\n");
        return 1;
    };
    int result[SIZE][SIZE];


    time_t first_time;
    time(&first_time);

    printf("Running Hill Climbing algorithm on %d threads...\n", NUM_THREADS);

    //Multithreading managment
    for (int i = 0; i < NUM_THREADS; i++) {
        memcpy(threadData[i].grid, initialGrid, sizeof(initialGrid));
        threadData[i].thread_id = i;

        pthread_create(&threads[i], NULL, thread, &threadData[i]);
    }
    for (int i = 0; i < NUM_THREADS; i++) pthread_join(threads[i], NULL);
    

    memcpy(result, threadData[solutionThreadId].grid, sizeof(result));


    time_t new_time;
    time(&new_time);


    //Print results
    printf("Found a solution :\n");
    printGrid(result);
    printf("Execution time: %d seconds.\n", (int)(new_time - first_time));

    return 0;
}